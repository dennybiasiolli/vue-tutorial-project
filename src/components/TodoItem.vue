<script setup>
import { ref, watch } from 'vue'

const props = defineProps({
  text: {
    type: String,
    required: true,
  },
  completed: {
    type: Boolean,
    required: true,
  }
})
const emit = defineEmits(['change', 'textChanged', 'delete'])

const showForm = ref(false)
const formValid = ref(false)
const innerText = ref('')

function handleChangeClick() {
  showForm.value = true
  innerText.value = props.text
}

function handleCancelClick() {
  showForm.value = false
}

function handleSave() {
  emit('textChanged', innerText.value)
  showForm.value = false
}

function validateForm() {
  if (!innerText.value.trim()) {
    formValid.value = false
    return
  }
  formValid.value = true
}

watch([innerText], () => {
  validateForm()
})
</script>

<template>
  <input
    type="checkbox"
    :checked="completed"
    @change="emit('change', $event.target.checked)"
  />
  {{ text }}
  <button v-if="!showForm" @click="handleChangeClick">Change</button>
  <button v-if="!showForm" @click="emit('delete')">Delete</button>
  <form v-show="showForm">
    <input type="text" v-model="innerText" />
    <button type="button" @click="handleCancelClick">Cancel</button>
    <button type="submit" :disabled="!formValid" @click="handleSave">Save</button>
  </form>
</template>
