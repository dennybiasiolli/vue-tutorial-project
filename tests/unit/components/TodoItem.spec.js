import { shallowMount } from '@vue/test-utils'
import { expect, it } from 'vitest'
import TodoItem from '@/components/TodoItem.vue'

it('should render as expected', () => {
  const wrapper = shallowMount(TodoItem, {
    props: {
      text: 'my todo text',
      completed: false,
    },
  })
  expect(wrapper.html()).toMatchSnapshot()
})

it('should emit change event', async () => {
  const wrapper = shallowMount(TodoItem, {
    props: {
      text: 'my todo text',
      completed: false,
    },
  })
  await wrapper.find('input').setChecked()
  expect(wrapper.emitted().change).toBeTruthy()
  expect(wrapper.emitted().change.length).toBe(1)
  expect(wrapper.emitted().change[0]).toEqual([true])
})
