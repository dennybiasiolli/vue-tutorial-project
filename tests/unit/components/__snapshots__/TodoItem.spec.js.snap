// Vitest Snapshot v1

exports[`should render as expected 1`] = `
"<input type=\\"checkbox\\">
 my todo text
<button>Change</button>
<button>Delete</button>
<form style=\\"display: none;\\"><input type=\\"text\\"><button type=\\"button\\">Cancel</button><button type=\\"submit\\" disabled=\\"\\">Save</button></form>"
`;
